osquery/logger/plugins/logger_plugin_util.h

```diff
@@ -0,0 +1,151 @@
+ protected:
+  static const std::chrono::seconds kLogPeriod;
+  static const size_t kMaxLogLines;
+
+ protected:
+  // These constructors are made available for subclasses to use, but
+  // subclasses should expose appropriate constructors to their users.
+  explicit BufferedLogForwarder(const std::string& name)
+      : log_period_(kLogPeriod),
+        max_log_lines_(kMaxLogLines),
+        log_index_(0),
+        index_name_(name) {}
+
+  template <class Rep, class Period>
+  explicit BufferedLogForwarder(
+      const std::string& name,
+      const std::chrono::duration<Rep, Period>& log_period)
+      : log_period_(
+            std::chrono::duration_cast<std::chrono::seconds>(log_period)),
+        max_log_lines_(kMaxLogLines),
+        log_index_(0),
+        index_name_(name) {}
+
+  template <class Rep, class Period>
+  explicit BufferedLogForwarder(
+      const std::string& name,
+      const std::chrono::duration<Rep, Period>& log_period,
+      size_t max_log_lines)
+      : log_period_(
+            std::chrono::duration_cast<std::chrono::seconds>(log_period)),
+        max_log_lines_(max_log_lines),
+        log_index_(0),
+        index_name_(name) {}
+
+ public:
+  /// A simple wait lock, and flush based on settings.
+  void start() override;
+
+  /**
+   * @brief Log a results string
+   *
+   * Writes the result string to the backing store for buffering, but *does
+   * not* actually send the string. The string will only be sent when check()
+   * runs and uses send() to send it.
+   *
+   * @param s Results string to log
+   */
+  Status logString(const std::string& s);
+
+  /**
+   * @brief Log a vector of status lines
+   *
+   * Decorates the status lines before writing to the backing store for
+   * buffering . *Does not* actually send the logs. The logs will only be sent
+   * when check() runs and uses send() to send them.
+   *
+   * @param log Vector of status lines to log
+   */
+  Status logStatus(const std::vector<StatusLogLine>& log);
+
+ protected:
+  /**
+   * @brief Send labeled result logs.
+   *
+   * The log_data provided to send must be mutable.
+   * To optimize for smaller memory, this will be moved into place within the
+   * constructed property tree before sending.
+   */
+  virtual Status send(std::vector<std::string>& log_data,
+                      const std::string& log_type) = 0;
+
+  /**
+   * @brief Check for new logs and send.
+   *
+   * Scan the logs domain for up to max_log_lines_ log lines.
+   * Sort those lines into status and request types then forward (send) each
+   * set. On success, clear the data and indexes.
+   */
+  void check();
+
+  /// Generate an index string to use with the backing store
+  std::string genLogIndex(bool results);
+
+  /// Seconds between flushing logs
+  std::chrono::seconds log_period_;
+
+  /// Max number of logs to flush per check
+  size_t max_log_lines_;
+
+  /// Hold an incrementing index for buffering logs
+  size_t log_index_;
```

Comment:
You can initialize this here:
```cpp
size_t log_index_{0};
```

Then you wont have to repeat the initialization in each constructor.